<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" >

    <!--base/world frame-->
    <link name="base_link"> </link>


    <!--Bottom base frame-->
    <joint name="B_base_joint" type="fixed">
        <origin xyz="0 0 0.05606" rpy="0 0 0"/>

        <parent link="base_link"/>
        <child link="B_base_link"/>        
    </joint>

    <link name="B_base_link">
        <visual>
            <origin xyz="0 0 -${B_base_length/2}" rpy="0 0 0"/>

            <geometry>    
                <!--<mesh filename="package://robot_pkg/meshes/base_link.STL" />-->
                <cylinder length="${B_base_length}" radius="${B_base_radius}"/>
            </geometry>

            <material name="white"/>
        </visual>

        <collision>
            <origin xyz="0 0 -${B_base_length/2}" rpy="0 0 0"/>

            <geometry>
                <!--<mesh filename="package://robot_pkg/meshes/base_link.STL" />-->            
                <cylinder length="${B_base_length}" radius="${B_base_radius}"/>
            </geometry>
        </collision>

        <xacro:inertial_cylinder mass="${B_base_mass}" length="${B_base_length}" radius="${B_base_radius}">
            <origin xyz="0 0 -${B_base_length/2}" rpy="0 0 0"/>

        </xacro:inertial_cylinder>
    </link>


    <!--Left wheel frame-->
    <joint name="L_wheel_joint" type="continuous">
        <origin xyz="0 -0.10575 -0.02356" rpy="${pi/2} ${pi/2} 0"/>

        <parent link="B_base_link"/>
        <child link="L_wheel_link"/>    
        <axis xyz="0 0 1"/>
        <limit effort="0.323" velocity="10.47" />    
    </joint>

    <link name="L_wheel_link">
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0"/>

            <geometry>    
                <cylinder length="${wheel_thickness}" radius="${wheel_radius}"/>
            </geometry>

            <material name="black"/>
        </visual>

        <collision>
            <origin xyz="0 0 0" rpy="0 0 0"/>

            <geometry>          
                <cylinder length="${wheel_thickness}" radius="${wheel_radius}"/>
            </geometry>
        </collision>

        <xacro:inertial_cylinder mass="${wheel_mass}" length="${wheel_thickness}" radius="${wheel_radius}">
            <origin xyz="0 0 0" rpy="0 0 0"/>

        </xacro:inertial_cylinder>
    </link>


    <!--Right wheel frame-->
    <joint name="R_wheel_joint" type="continuous">
        <origin xyz="0 0.10575 -0.02356" rpy="${-pi/2} ${pi/2} 0"/>

        <parent link="B_base_link"/>
        <child link="R_wheel_link"/>    
        <axis xyz="0 0 -1"/>
        <limit effort="0.323" velocity="10.47" />    
    </joint>

    <link name="R_wheel_link">
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0"/>

            <geometry>    
                <cylinder length="${wheel_thickness}" radius="${wheel_radius}"/>
            </geometry>

            <material name="black"/>
        </visual>

        <collision>
            <origin xyz="0 0 0" rpy="0 0 0"/>

            <geometry>          
                <cylinder length="${wheel_thickness}" radius="${wheel_radius}"/>
            </geometry>
        </collision>

        <xacro:inertial_cylinder mass="${wheel_mass}" length="${wheel_thickness}" radius="${wheel_radius}">
            <origin xyz="0 0 0" rpy="0 0 0"/>

        </xacro:inertial_cylinder>
    </link>


    <!--Caster wheel base 1-->
    <joint name="U_base_wheel_1_joint" type="continuous">
        <origin xyz="0.05751 0 -0.0278" rpy="${pi} 0 0"/>

        <parent link="B_base_link"/>
        <child link="U_base_wheel_1_link"/>    
        <axis xyz="0 0 1"/>
    </joint>

    <link name="U_base_wheel_1_link">
        <visual>
            <origin xyz="0 0 -0.00329" rpy="0 0 0"/>

            <geometry>    
                <box size="${U_base_box_X} ${U_base_box_Y} ${U_base_box_Z}"/>
            </geometry>

            <material name="light_gray"/>
        </visual>

        <collision>
            <origin xyz="0 0 -0.00329" rpy="0 0 0"/>

            <geometry>    
                <box size="${U_base_box_X} ${U_base_box_Y} ${U_base_box_Z}"/>
            </geometry>
        </collision>

        <xacro:inertial_box mass="${U_base_mass}" x="${U_base_box_X}" y="${U_base_box_Y}" z="${U_base_box_Z}">
            <origin xyz="0 0 -0.00329" rpy="0 0 0"/>

        </xacro:inertial_box>
    </link>


    <!--Caster wheel 1-->
    <joint name="U_wheel_1_joint" type="continuous">
        <origin xyz="0.01577 0 0.01622" rpy="${pi/2} 0 0"/>

        <parent link="U_base_wheel_1_link"/>
        <child link="U_wheel_1_link"/>    
        <axis xyz="0 0 1"/>   
    </joint>

    <link name="U_wheel_1_link">
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0"/>

            <geometry>    
                <cylinder radius="${U_wheel_radius}" length="${U_wheel_thickness}"/>
            </geometry>

            <material name="white"/>
        </visual>

        <collision>
            <origin xyz="0 0 0" rpy="0 0 0"/>

            <geometry>    
                <cylinder radius="${U_wheel_radius}" length="${U_wheel_thickness}"/>
            </geometry>
        </collision>

        <xacro:inertial_cylinder mass="${U_wheel_mass}" length="${U_wheel_thickness}" radius="${U_wheel_radius}">
            <origin xyz="0 0 0" rpy="0 0 0"/>
        
        </xacro:inertial_cylinder>
    </link>


    <!--Caster wheel base 2-->
    <joint name="U_base_wheel_2_joint" type="continuous">
        <origin xyz="-0.05751 0 -0.0278" rpy="${pi} 0 ${pi}"/>

        <parent link="B_base_link"/>
        <child link="U_base_wheel_2_link"/>    
        <axis xyz="0 0 1"/>
    </joint>

    <link name="U_base_wheel_2_link">
        <visual>
            <origin xyz="0 0 -0.00329" rpy="0 0 0"/>

            <geometry>    
                <box size="${U_base_box_X} ${U_base_box_Y} ${U_base_box_Z}"/>
            </geometry>

            <material name="light_gray"/>
        </visual>

        <collision>
            <origin xyz="0 0 -0.00329" rpy="0 0 0"/>

            <geometry>    
                <box size="${U_base_box_X} ${U_base_box_Y} ${U_base_box_Z}"/>
            </geometry>
        </collision>

        <xacro:inertial_box mass="${U_base_mass}" x="${U_base_box_X}" y="${U_base_box_Y}" z="${U_base_box_Z}">
            <origin xyz="0 0 -0.00329" rpy="0 0 0"/>

        </xacro:inertial_box>
    </link>


    <!--Caster wheel 2-->
    <joint name="U_wheel_2_joint" type="continuous">
        <origin xyz="0.01577 0 0.01622" rpy="${pi/2} 0 0"/>

        <parent link="U_base_wheel_2_link"/>
        <child link="U_wheel_2_link"/>    
        <axis xyz="0 0 -1"/>  
    </joint>

    <link name="U_wheel_2_link">
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0"/>

            <geometry>    
                <cylinder radius="${U_wheel_radius}" length="${U_wheel_thickness}"/>
            </geometry>

            <material name="white"/>
        </visual>

        <collision>
            <origin xyz="0 0 0" rpy="0 0 0"/>

            <geometry>    
                <cylinder radius="${U_wheel_radius}" length="${U_wheel_thickness}"/>
            </geometry>
        </collision>

        <xacro:inertial_cylinder mass="${U_wheel_mass}" length="${U_wheel_thickness}" radius="${U_wheel_radius}">
            <origin xyz="0 0 0" rpy="0 0 0"/>
        
        </xacro:inertial_cylinder>
    </link>


    <!--Fixed support-->
    <joint name="Support_joint" type="fixed">
        <origin xyz="0 0 0.101" rpy="0 0 0"/>

        <parent link="B_base_link"/>
        <child link="Support_link"/>        
    </joint>

    <link name="Support_link">
        <visual>
            <origin xyz="0 0 -${Support_length/2}" rpy="0 0 0"/>

            <geometry>    
                <cylinder length="${Support_length}" radius="${Support_radius}"/>
            </geometry>

            <material name="white"/>
        </visual>

        <collision>
            <origin xyz="0 0 -${Support_length/2}" rpy="0 0 0"/>

            <geometry>          
                <cylinder length="${Support_length}" radius="${Support_radius}"/>
            </geometry>
        </collision>

        <xacro:inertial_cylinder mass="${Support_mass}" length="${Support_length}" radius="${Support_radius}">
            <origin xyz="0 0 -${Support_length/2}" rpy="0 0 0"/>

        </xacro:inertial_cylinder>
    </link>

</robot>